core/finality-grandpa/src/import.rs

```diff
@@ -425,7 +425,7 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA, SC> BlockImport<Block>
 
 		// we don't want to finalize on `inner.import_block`
 		let mut justification = block.justification.take();
-		let enacts_consensus_change = new_cache.contains_key(&well_known_cache_keys::AUTHORITIES);
+		let enacts_consensus_change = !new_cache.is_empty();
```

Comment:
It **will** take it into account - please see how `enacts_consensus_change` is processed: when it's true, it'll fire `self.consensus_changes.lock().note_change((number, hash))` and then this will be processed by the `environment/finalize_block.rs` - see how `justification_required` is built. So the justification will be emitted for such blocks.

But OK - if you do not consider epoch change as consensus change, then let's not emit justification in this case.

Reply:
Yes, but the problem is that the GRANDPA voters currently won't take that into account, and will not limit their votes to such blocks (like they do for authority changes), so we might not get a justification for it (we just finalize a further block for example).
@rphmeier Can you comment on whether epoch change should be considered a consensus change?

Reply:
The only downside I see is that we'll get more justifications. Although the point is kind of moot in general because the GRANDPA authorities will change 99% of the time on the same block as BABE or other ones do. It's unlikely that they'll be out of sync.